test/extended/util/ldap.go

```diff
@@ -0,0 +1,310 @@
+		ObjectMeta: v1.ObjectMeta{
+			Name: saName,
+		},
+	}); err != nil {
+		return "", nil, err
+	}
+
+	// Create CA.
+	ca, err := crypto.MakeSelfSignedCA(path.Join(certDir, caCertFilename), path.Join(certDir, caKeyFilename),
+		path.Join(certDir, "serial"), caName, 100)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Ensure that the server cert is valid for localhost and the service network hostname.
+	serviceHost := ldapService.Name + "." + oc.Namespace() + ".svc"
+	serverCertConfig, err := ca.MakeServerCert(sets.NewString("localhost", "127.0.0.1", serviceHost), 100)
+	if err != nil {
+		return "", nil, err
+	}
+
+	caPEM, _, err := ca.Config.GetPEMBytes()
+	if err != nil {
+		return "", nil, err
+	}
+
+	serverCertPEM, serverCertKeyPEM, err := serverCertConfig.GetPEMBytes()
+	if err != nil {
+		return "", nil, err
+	}
+
+	_, err = oc.AdminKubeClient().CoreV1().Secrets(oc.Namespace()).Create(&corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name: certMountName,
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       []byte(serverCertPEM),
+			corev1.TLSPrivateKeyKey: serverCertKeyPEM,
+			caCertFilename:          caPEM,
+		},
+		Type: corev1.SecretTypeTLS,
+	})
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Allow the openldap container to run as root and privileged. This lets us use the existing openldap server
+	// container startup scripts to poplate its database using ldapi:///.
+	// TODO: Turn these, (and other resources in this function) into yamls.
+	err = oc.AsAdmin().Run("create").Args("role", "scc-anyuid", "--verb=use", "--resource=scc",
+		"--resource-name=anyuid").Execute()
+	if err != nil {
+		return "", nil, err
+	}
+	err = oc.AsAdmin().Run("adm").Args("policy", "add-role-to-user", "scc-anyuid", "-z", "ldap",
+		"--role-namespace", oc.Namespace()).Execute()
+	if err != nil {
+		return "", nil, err
+	}
+
+	err = oc.AsAdmin().Run("create").Args("role", "scc-priv", "--verb=use", "--resource=scc",
+		"--resource-name=privileged").Execute()
+	if err != nil {
+		return "", nil, err
+	}
+	err = oc.AsAdmin().Run("adm").Args("policy", "add-role-to-user", "scc-priv", "-z", "ldap",
+		"--role-namespace", oc.Namespace()).Execute()
+	if err != nil {
+		return "", nil, err
+	}
+
+	serverDeployment, err := oc.AdminKubeClient().AppsV1().Deployments(oc.Namespace()).Create(deploy)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Wait for an available replica.
+	err = wait.PollImmediate(1*time.Second, 5*time.Minute, func() (done bool, err error) {
+		dep, getErr := oc.AdminKubeClient().AppsV1().Deployments(oc.Namespace()).Get(serverDeployment.Name,
+			v1.GetOptions{})
+		if getErr != nil {
+			return false, getErr
+		}
+		if dep.Status.AvailableReplicas == 0 {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return "", nil, fmt.Errorf("replica for %s not avaiable: %v", serverDeployment.Name, err)
+	}
+
+	// Confirm ldap server availability.
+	if err := checkLDAPConn(oc, serviceHost); err != nil {
```

Comment:
If I remember correctly, you had issues with routes, which is why you're doing this over the internal network. Could you please mention that in the comment somewhere? Otherwise people might be confused why aren't you just checking from the external network.